scripts/lib/api/generateApiComponents.ts

```diff
@@ -0,0 +1,138 @@
+// Generate MDX components
+// ------------------------------------------------------------------
+
+/**
+ * Creates the opening tag of the API components. The function sets all possible
+ * props values even if they are empty or undefined. All the props without value
+ * will be removed when generating the markdown file in `htmlToMd.ts`.
+ */
+export async function createOpeningTag(
+  tagName: string,
+  props: ComponentProps,
+): Promise<string> {
+  const type = props.attributeTypeHint?.replaceAll("'", APOSTROPHE_HEX_CODE);
+  const value = props.attributeValue?.replaceAll("'", APOSTROPHE_HEX_CODE);
+  const signature = await htmlSignatureToMd(props.rawSignature!);
+  const extraSignatures: string[] = [];
+  for (const sig of props.extraRawSignatures ?? []) {
+    extraSignatures.push(
+      `${APOSTROPHE_HEX_CODE}${await htmlSignatureToMd(
+        sig!,
+      )}${APOSTROPHE_HEX_CODE}`,
+    );
+  }
+
+  return `<${tagName} 
+    id='${props.id}'
+    name='${props.name}'
+    attributeTypeHint='${type}'
+    attributeValue='${value}'
+    github='${props.githubSourceLink}'
+    signature='${signature}'
+    extraSignatures='[${extraSignatures.join(", ")}]'
+    >
+  `;
+}
+
+/**
+ * Removes the original link from sphinx.ext.viewcode and returns the HTML string for our own link.
+ *
+ * This returns the HTML string, rather than directly inserting into the HTML, because the insertion
+ * logic is most easily handled by the calling code.
+ *
+ * This function works the same regardless of whether the Sphinx build used `sphinx.ext.viewcode`
+ * or `sphinx.ext.linkcode` because they have the same HTML structure.
+ *
+ * If the link corresponds to a method, we only return a link if it has line numbers included,
+ * which implies that the link came from `sphinx.ext.linkcode` rather than `sphinx.ext.viewcode`.
+ * That's because the owning class will already have a link to the relevant file; it's
+ * noisy and not helpful to repeat the same link without line numbers for the individual methods.
+ */
+export function prepareGitHubLink(
+  $child: Cheerio<any>,
+  isMethod: boolean,
+): string | undefined {
+  const originalLink = $child.find(".viewcode-link").closest("a");
+  if (originalLink.length === 0) {
+    return undefined;
+  }
+  const href = originalLink.attr("href")!;
+  originalLink.first().remove();
+  return !isMethod || href.includes(".py#") ? href : undefined;
+}
+
+/**
+ * Find the element that both matches the `selector` and whose content is the same as `text`
+ */
+function findByText(
+  $: CheerioAPI,
+  $main: Cheerio<any>,
+  selector: string,
+  text: string,
+): Cheerio<any> {
+  return $main.find(selector).filter((i, el) => $(el).text().trim() === text);
+}
+
+export function addExtraSignatures(
+  componentProps: ComponentProps,
+  extraRawSignatures: ComponentProps[],
+): void {
+  componentProps.extraRawSignatures = [
+    ...extraRawSignatures.flatMap((sigProps) => sigProps.rawSignature ?? []),
+  ];
```

Comment:
Using `flatMap` surprises me because `rawSignature` is a string. Should this be

```
componentProps.extraRawSignatures = Array.from(
    extraRawSignatures.map((x) => x.signature).filter((sig) => sig !== undefined)
)
```

Also can `extraRawSignatures` have a more precise type like `string[]` or `Array<string | undefined>`? Seems weird to have all the ComponentProps if we're going to discard them.

Reply:
In reality, `rawSignature` is a `string` but could be `undefined` given that we have API types without it. The problem with using the `filter` is that we get the type `(string | undefined)[]` instead of `string[] | undefined` as in the `flatMap`.

> Also can `extraRawSignatures` have a more precise type 

This is a good idea. I think we could remove the `flatMap` or the `map` + `filter` entirely, but I would like to do it in #1026 as a follow-up because we can simplify the main function by only storing the overloaded signature and not all props. The function `addExtraSignatures` will make all the work as a black box